test/lib/Elastica/Test/Aggregation/CardinalityTest.php

```diff
@@ -33,4 +33,92 @@ public function testCardinalityAggregation()
 
         $this->assertEquals(3, $results['value']);
     }
+
+    /**
+     * @dataProvider invalidPrecisionThresholdProvider
+     * @expectedException \InvalidArgumentException
+     * @param $threshold
+     */
+    public function testInvalidPrecisionThreshold($threshold)
+    {
+        $agg = new Cardinality('threshold');
+        $agg->setPrecisionThreshold($threshold);
+    }
+
+    /**
+     * @dataProvider validPrecisionThresholdProvider
+     * @param $threshold
+     */
+    public function testPrecisionThreshold($threshold)
+    {
+        $agg = new Cardinality('threshold');
+        $agg->setPrecisionThreshold($threshold);
+
+        $this->assertNotNull($agg->getParam('precision_threshold'));
+        $this->assertInternalType('int', $agg->getParam('precision_threshold'));
+    }
+
+    public function invalidPrecisionThresholdProvider()
+    {
+        return array(
+            'string' => array('100'),
+            'float' => array(7.8),
+            'boolean' => array(true),
+            'array' => array(array()),
+            'object' => array(new \StdClass),
+        );
+    }
+
+    public function validPrecisionThresholdProvider()
+    {
+        return array(
+            'negative-int' => array(-140),
+            'zero' => array(0),
+            'positive-int' => array(150),
+            'more-than-max' => array(40001),
+        );
+    }
+
+    /**
+     * @dataProvider validRehashProvider
+     * @param bool $rehash
+     */
+    public function testRehash($rehash)
+    {
+        $agg = new Cardinality('rehash');
+        $agg->setRehash($rehash);
+
+        $this->assertNotNull($agg->getParam('rehash'));
+        $this->assertInternalType('boolean', $agg->getParam('rehash'));
+    }
+
+    /**
+     * @dataProvider invalidRehashProvider
+     * @expectedException \InvalidArgumentException
+     * @param mixed $rehash
+     */
+    public function testInvalidRehash($rehash)
+    {
+        $agg = new Cardinality('rehash');
+        $agg->setRehash($rehash);
+    }
+
+    public function invalidRehashProvider()
```

Comment:
Thanks for the quick review @ruflin! The annotated data providers need to be public so they can be called using reflection by PHPUnit.
https://phpunit.de/manual/4.6/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers